### PR TITLE
vcmi: build lua with STRICT_COMPILATION

### PR DESCRIPTION
**Pull request by rilian-la-te**
This is a fixes for some LUA warnings for building it with strict compilation.

Let's eat modernizing elephant by pieces) 

Also modernize LUA by clang-tidy, as it is harmless AFAIK.

### CODE REVIEW

File: scripting/lua/LuaScriptingContext.cpp [hunk 1]
```diff
@@ -206,7 +205,7 @@ JsonNode LuaContext::callGlobal(const std::string & name, const JsonNode & param
 
 		S.clear();
 
-		return JsonNode();
+		return {};
```
Comment (dydzio0614) [inline scripting/lua/LuaScriptingContext.cpp:208]: I personally like previous version more

Comment (rilian-la-te) [reply scripting/lua/LuaScriptingContext.cpp:208]: @dydzio0614 Will look for check which cause this behaviour.

Comment (rilian-la-te) [reply scripting/lua/LuaScriptingContext.cpp]: @dydzio0614 removed modernization

**Comment by rilian-la-te** [comment]
@IvanSavenko can you review? @dydzio0614 told me than I should ask you about a review)

File: scripting/lua/api/netpacks/BattleUnitsChanged.cpp [hunk 2]
```diff
@@ -44,7 +44,7 @@ int BattleUnitsChangedProxy::add(lua_State * L)
 	if(!S.tryGet(1, object))
 		return S.retVoid();
 
-	uint32_t id;
+	uint32_t id = 0;
 
 	if(!S.tryGet(2, id))
 		return S.retVoid();
```
Comment (kambala-decapitator) [inline scripting/lua/api/netpacks/BattleUnitsChanged.cpp:50]: I think this is superfluous. If I read the logic right: if `id` can't be filled from Lua, return immediately.

File: scripting/lua/api/netpacks/BattleUnitsChanged.cpp [hunk 3]
```diff
@@ -70,7 +70,7 @@ int BattleUnitsChangedProxy::update(lua_State * L)
 	if(!S.tryGet(1, object))
 		return S.retVoid();
 
-	uint32_t id;
+	uint32_t id = 0;
```
Comment (kambala-decapitator) [inline scripting/lua/api/netpacks/BattleUnitsChanged.cpp:73]: same here

**Review by kambala-decapitator — APPROVED** [review]

File: scripting/lua/api/netpacks/BattleUnitsChanged.cpp (hunk 2 above)
Comment (rilian-la-te) [reply scripting/lua/api/netpacks/BattleUnitsChanged.cpp:50]: @kambala-decapitator this is filled because clang-tidy initialized all variables. I can block this with //NOLINT, but do I need to block it?

File: scripting/lua/api/netpacks/BattleUnitsChanged.cpp (hunk 3 above)
Comment (rilian-la-te) [reply scripting/lua/api/netpacks/BattleUnitsChanged.cpp:73]: same as above.

File: scripting/lua/LuaCallWrapper.h [hunk 4]
```diff
@@ -196,7 +197,7 @@ class LuaMethodWrapper <U, R(T:: *)(P1)const, method>
 		if(!S.tryGet(1,obj))
 			return S.retVoid();
 
-		P1 p1;
+		std::remove_cv_t<std::remove_reference_t<P1>> p1;
```
Comment (IvanSavenko) [inline scripting/lua/LuaCallWrapper.h:200]: Why is this needed?

File: scripting/lua/LuaReference.cpp [hunk 5]
```diff
@@ -23,7 +23,7 @@ LuaReference::LuaReference(lua_State * L)
 	key = luaL_ref(l, LUA_REGISTRYINDEX);
 }
 
-LuaReference::LuaReference(LuaReference && other)
+LuaReference::LuaReference(LuaReference && other) noexcept
```
Comment (IvanSavenko) [inline scripting/lua/LuaReference.cpp:26]: Do we use `noexcept`? Is there any point in using it?
Not against this change but what are advantages of this form?

File: scripting/lua/LuaScriptModule.cpp [hunk 6]
```diff
@@ -17,7 +17,7 @@
 #define strcpy_s(a, b, c) strncpy(a, c, b)
 #endif
 
-const char *g_cszAiName = "Lua interpreter";
+static const char *g_cszAiName = "Lua interpreter";
```
Comment (IvanSavenko) [inline scripting/lua/LuaScriptModule.cpp:20]: Frankly, the only useful change I noticed so far

File: scripting/lua/api/BonusSystem.h [hunk 7]
```diff
@@ -54,7 +54,7 @@ class BonusListProxy : public SharedWrapper<const BonusList, BonusListProxy>
 
 	static const std::vector<typename Wrapper::CustomRegType> REGISTER_CUSTOM;
 
-	static std::shared_ptr<const Bonus> index(std::shared_ptr<const BonusList> self, int key);
+	static std::shared_ptr<const Bonus> index(std::shared_ptr<const BonusList> self, int key); //NOLINT
```
Comment (IvanSavenko) [inline scripting/lua/api/BonusSystem.h:57]: Again, comments that are not comment but attributes...

File: scripting/lua/api/netpacks/BattleUnitsChanged.cpp (hunk 2 above)
Comment (IvanSavenko) [reply scripting/lua/api/netpacks/BattleUnitsChanged.cpp:50]: Then don't use it? Or significantly clean up its output?
So far I don't see any significant improvements neither here nor in #1559

**Review by IvanSavenko — COMMENTED** [review]
Don't see any significant improvements to be honest.
Not *against* these changes, but they are mostly redundant and have zero need to allocate time on.

File: scripting/lua/LuaCallWrapper.h (hunk 4 above)
Comment (rilian-la-te) [reply scripting/lua/LuaCallWrapper.h:200]: @kambala-decapitator In library series some types in VCMI library is converted from passing by value to passing by a const reference. If this change is not implemented, then Lua will breaks on const references passed in this templates. 

This change will strip constness and reference from such types, and just generate a new temporary value.

File: scripting/lua/LuaReference.cpp (hunk 5 above)
Comment (rilian-la-te) [reply scripting/lua/LuaReference.cpp:26]: @IvanSavenko standard library do not call move constructors if they are not marked noexcept, which leads to copy a variable everywhere. Maybe it this case this is not significant.

File: scripting/lua/LuaScriptModule.cpp (hunk 6 above)
Comment (rilian-la-te) [reply scripting/lua/LuaScriptModule.cpp:20]: This change is not from modernization, I done it by hand, because Lua do not builds without it)

File: scripting/lua/LuaCallWrapper.h (hunk 4 above)
Comment (IvanSavenko) [reply scripting/lua/LuaCallWrapper.h:200]: To clarify - this is actual behavior that *is* happening and not yet another "improvement" from clang-tidy?

In that case perhaps better to add using at the very start of the class:
`using Parameter1 = td::remove_cv_t<std::remove_reference_t<P1>>;`
And then declare variables of type `Parameter1`

Comment (rilian-la-te) [reply scripting/lua/LuaCallWrapper.h:200]: @IvanSavenko This is not another improvement from clang-tidy. But it will noticed only if you pass const reference to this templates.

Will do your suggestion.

File: scripting/lua/api/netpacks/BattleUnitsChanged.cpp (hunk 2 above)
Comment (rilian-la-te) [reply scripting/lua/api/netpacks/BattleUnitsChanged.cpp:50]: @IvanSavenko some constructors and destructors made empty by clang-tidy itself and got defaulted by next passes. Human would not detect it as is, if he will not review a code in details.

Comment (IvanSavenko) [reply scripting/lua/api/netpacks/BattleUnitsChanged.cpp:50]: Example? Most complex case I saw is `ObstacleSideOptions::ObstacleSideOptions` [here](https://github.com/vcmi/vcmi/pull/1559/files#diff-b371b0903dfc2aa65381f853eef4f160dd6e624a48edc4ca095550124bcc9d27).
Still clearly an empty constructor that only has default-init of member and parent class.

Nice, but not something worth spending time on.

Comment (rilian-la-te) [reply scripting/lua/api/netpacks/BattleUnitsChanged.cpp:50]: @IvanSavenko here is an example of defaulting a huge copy constructor from https://github.com/vcmi/vcmi/pull/1554/commits/0b22bf0f3f51ee0e102c2effaddb21ffe1b96c88#diff-7906420fb6fbc1c2a5f88c4dc055c77fa91e77aadf5b5f56dda376e2e85a2323

This is from first attempt of clang-tidy, but second attempt will do the same.

Comment (IvanSavenko) [reply scripting/lua/api/netpacks/BattleUnitsChanged.cpp:50]: Nice, although seems to be the only case, and poorly written even from c++03 perspective.
Would be useful later, for c++20 default comparisons though.

Comment (rilian-la-te) [reply scripting/lua/api/netpacks/BattleUnitsChanged.cpp:50]: @IvanSavenko I can remove checks you do not like, like init-all-variables and do modernization again)

Comment (rilian-la-te) [reply scripting/lua/api/netpacks/BattleUnitsChanged.cpp]: @IvanSavenko cppcoreguidelines-init-all removed

File: scripting/lua/api/BonusSystem.h (hunk 7 above)
Comment (rilian-la-te) [reply scripting/lua/api/BonusSystem.h]: Removed temporarily.

File: scripting/lua/LuaCallWrapper.h (hunk 4 above)
Comment (rilian-la-te) [reply scripting/lua/LuaCallWrapper.h]: @IvanSavenko your suggestion done for using.

File: scripting/lua/api/netpacks/BattleUnitsChanged.cpp (hunk 3 above)
Comment (rilian-la-te) [reply scripting/lua/api/netpacks/BattleUnitsChanged.cpp]: removed

**Comment by kambala-decapitator** [comment]
looks fine to me, let's wait for @IvanSavenko's feedback

File: scripting/lua/LuaScriptingContext.cpp [hunk 8]
```diff
@@ -35,12 +35,8 @@ namespace scripting
 const std::string LuaContext::STATE_FIELD = "DATA";
 
 LuaContext::LuaContext(const Script * source, const Environment * env_)
-	: ContextBase(env_->logger()),
-	script(source),
-	env(env_)
+	: ContextBase(env_->logger()), L(luaL_newstate()), script(source), env(env_)
```
[the remark this answers is not in the document]
Comment (rilian-la-te) [reply scripting/lua/LuaScriptingContext.cpp]: done

File: scripting/lua/LuaStack.cpp [hunk 9]
```diff
@@ -20,9 +20,8 @@ namespace scripting
 {
 
 LuaStack::LuaStack(lua_State * L_)
-	: L(L_)
+	: L(L_), initialTop(lua_gettop(L))
```
[the remark this answers is not in the document]
Comment (rilian-la-te) [reply scripting/lua/LuaStack.cpp]: done

**Comment by rilian-la-te** [comment]
@IvanSavenko waiting for merge this and #1559 )
